dowhy/causal_refuters/dummy_outcome_refuter.py

```diff
@@ -105,30 +105,41 @@ def refute_estimate(self):
         self.logger.info("Refutation over {} simulated datasets".format(self._num_simulations) )
         self.logger.info("The transformation passed: {}".format(self._transformations) )
 
-        data_chunks = self.preprocess_data_by_treatment()
+        groups = self.preprocess_data_by_treatment()
         estimates = []
-        for chunk in data_chunks:
-
-            X_input = chunk[self._chosen_variables]
-            new_outcome = chunk['y']
-            X = self._data[self._chosen_variables]
+        for key_train, _ in groups:
+            X_train = groups.get_group(key_train)[self._chosen_variables].values
+            new_outcome_train = groups.get_group(key_train)['y'].values
+            validation_df = [] 
+            for key_validation, _ in groups:
+                if key_validation != key_train:
+                    validation_df.append(groups.get_group(key_validation))
+
+            validation_df = pd.concat(validation_df)
+            X_validation = validation_df[self._chosen_variables].values
+            new_outcome_validation = validation_df['y'].values
 

```

Comment:
Before the loop, (Even before creating groups), you need to check if the pipeline only contains permute, noise and zero. If so, then you don't need to create a validationdf. You can use the full data. But if there is at least one action that is outside these three, then you need a validation set.

So one way is to check for these three on line 108 and create groups only if there are functions included in the pipeline. Then using the code in point 2) in the comment above, you can either run a loop over k keys.

If no such function is include, simply process the pipeline once, as you were doing before we started thinking of ML.

To avoid code-repetition, you can wrap the "for action, func_args" loop in a new method, called process_pipeline(train, validate). so that you can call it easily two times based on the if condition in the above paragraph. When the functions are not present in the pipeline, you can simply pass (train=None, validate=full_data) and it should work.

Does this make sense?

Reply:
Yeah, it makes sense. I've added a flag ```no_estimator``` to account for this possibility of no estimators in the list of transformations. Furthermore, I have added the function ```process_data``` to separately process the pipeline.